### PR TITLE
Features/types

### PR DESCRIPTION
**Pull request by Markus-Goetz**
Realized HeAT's type system. Enabled type casting. Major refactoring of existing code to utilize new type system.

Resolves issues #10 and #12

Detailed overview of the changes:

* setup.py requires torch and numpy, h5py is optional dependency
* added io module
    - contains optional hdf5 loading function
    - enhanced hdf5 loading function to support arbitrary dimensions
    - added documentation
    - added tests
* added random module
    - copy over of the existing global seed initialization and uniform sampling
    - test cases and documentation missing
* enhanced communicator module
    - introduced abstract Communicator base class
    - communicators must now always be instantiated for a tensor
    - each communicator now has a chunk method, i.e. dividing a shape into slices according to the communication strategy and split axis, previously part of the load_h5 method
    - added tests
    - added documentation
* removed float16/half type, only rudimentary CPU support from PyTorch
* core module pulls io unqualified, random qualified, tensor qualified through to the parent module
* enhanced tensor module
    - introduced new constructor method
    - refactored existing code to utilize existing constructor
    - added a number of TODOs for missing documentation, tests, functionality, ...
    - added tested and documented astype cast function (see issue #10)
* refactored kmeans code
    - now utilizes the refactored random and tensor modules
    - refactored tests cases to utilize Python's unittest, need to be enhanced to better check KMeans functionality
    - added TODOs for missing documentation and functionality

### CODE REVIEW

File: heat/core/tests/test_types.py [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+        self.assert_is_heat_type(heat_type)
+
+        # check a type constructor without any value
+        no_value = heat_type()
+        self.assertIsInstance(no_value, ht.tensor)
+        self.assertEqual(no_value.shape, (1,))
+        self.assertEqual((no_value._tensor__array == 0).all().item(), 1)
+        self.assertEqual(no_value._tensor__array.dtype, torch_type)
+
+        # check a type constructor with a complex value
```
Comment (d1saster) [inline heat/core/tests/test_types.py]: Complex value reminds me too much of a complex number. Wouldn't something like "elaborate input array" fit better, or just `elaborate_array` in order to have a short variable name?

File: heat/core/types.py [hunk 2]
```diff
@@ -0,0 +1,257 @@
...
+        try:
+            array = torch.tensor(*value, dtype=torch_type)
+        except TypeError as exception:
+            # re-raise the exception to be consistent with numpy's exception interface
+            raise ValueError(str(exception))
+
+        return tensor.tensor(array, tuple(array.shape), split=None, comm=NoneCommunicator())
+
+    @classmethod
+    @abc.abstractclassmethod
```
Comment (d1saster) [inline heat/core/types.py]: `abc.abstractclassmethod` has become deprecated in Python 3.3, according to the [documentation ](https://docs.python.org/3/library/abc.html#abc.abstractclassmethod). `abc.abstractmethod` should be sufficient.

File: heat/core/communicator.py [hunk 3]
```diff
@@ -1,25 +1,91 @@
...
-class NoneCommunicator:
+
+__all__ = [
+    'mpi'
+]
+
+
+class Communicator(metaclass=abc.ABCMeta):
+    @staticmethod
+    @abc.abstractstaticmethod
```
Comment (d1saster) [inline heat/core/communicator.py]: `abc.abstractstaticmethod` has become deprecated in Python 3.3, according to the [documentation ](https://docs.python.org/3/library/abc.html#abc.abstractstaticmethod). `abc.abstractmethod` should be sufficient. Btw. is there some special python 3 version we are targeting? Pytorch seems to need at least 3.5 if one chooses python3.
